syntax = "proto3";

service CoteaGateway {
   rpc StartSession(EmptyMsg) returns (StartSessionMSG) {}
   rpc InitExecution(Config) returns (Status) {}
   rpc RunTask(Task) returns (TaskResults) {}
   rpc StopExecution(SessionID) returns (Status) {}
   // rpc RestartExecution(SessionID) returns (Status) {}
}

service CoteaWorker {
   rpc InitExecution(WorkerConfig) returns (Status) {}
   rpc RunTask(WorkerTask) returns (TaskResults) {}
   rpc StopExecution(EmptyMsg) returns (Status) {}
   rpc HealthCheck(EmptyMsg) returns (WorkerHealthStatus) {}
   // rpc RestartExecution() returns (Status) {}
}

message StartSessionMSG {
   bool ok = 1;
   string ID = 2;
   string error_msg = 3;
}

message SessionID {
   string session_ID = 1;
}

message EmptyMsg {}

message MapFieldEntry {
   string key = 1;
   string value = 2;
} 

message Config {
   string session_ID = 1;
   string hosts = 2;
   string inv_path = 3;
   string extra_vars = 4;
   repeated MapFieldEntry env_vars = 5;
   string ansible_library = 6;
   bool not_gather_facts = 7;
}

message WorkerConfig {
   string hosts = 1;
   string inv_path = 2;
   string extra_vars = 3;
   repeated MapFieldEntry env_vars = 4;
   string ansible_library = 5;
   bool not_gather_facts = 6;
}

message Task {
   string session_ID = 1;
   string task_str = 2;
   bool is_dict = 3;
}

message WorkerTask {
   string task_str = 1;
   bool is_dict = 2;
}

message TaskResult {
   bool ok = 1;
   string results_dict_str = 2;
   string task_name = 3;
   bool is_changed = 4;
   bool is_failed = 5;
   bool is_skipped = 6;
   bool is_unreachable = 7;
   bool is_ignored_errors = 8;
   bool is_ignored_unreachable = 9;
   string stdout = 10;
   string stderr = 11;
   string msg = 12;
}

message TaskResults {
   bool task_adding_ok = 1;
   string task_adding_error = 2;
   repeated TaskResult task_results = 3;
}

message Status {
   bool ok = 1;
   string error_msg = 2;
}

message WorkerHealthStatus {
   bool ok = 1;
   int32 executions_count = 2;
   int32 executed_tasks_count = 3;
}
